src/common/walFilter/walFilter.c

```diff
@@ -0,0 +1,722 @@
+    this->gotLen = 0;
+}
+
+static const StorageRead *
+getNearWal (WalFilterState *const this, bool isNext)
+{
+    const String *walSegment = NULL;
+    const TimeLineID timeLine = this->currentHeader->xlp_tli;
+    uint64_t segno = this->currentHeader->xlp_pageaddr / this->segSize;
+    const String *const path = strNewFmt("%08X%08X", timeLine, (uint32) (segno / XLogSegmentsPerXLogId(this->segSize)));
+
+    const StringList *const segmentList = storageListP(
+        storageRepoIdx(this->archiveInfo->repoIdx),
+        strNewFmt(STORAGE_REPO_ARCHIVE "/%s/%s", strZ(this->archiveInfo->archiveId), strZ(path)),
+        .expression = strNewFmt("^[0-f]{24}-[0-f]{40}" COMPRESS_TYPE_REGEXP "{0,1}$"));
+
+    if (strLstEmpty(segmentList))
+    {
+        // an exotic case where we couldn't even find the current file.
+        THROW(FormatError, "no WAL files were found in the repository");
+    }
+
+    uint64_t segnoDiff = UINT64_MAX;
+    for (uint32_t i = 0; i < strLstSize(segmentList); i++)
+    {
+        const String *const file = strSubN(strLstGet(segmentList, i), 0, 24);
+        TimeLineID tli;
+        uint64_t fileSegNo = 0;
+        XLogFromFileName(strZ(file), &tli, &fileSegNo, this->segSize);
+
+        if (isNext)
+        {
+            if (fileSegNo - segno < segnoDiff && fileSegNo > segno)
+            {
+                segnoDiff = fileSegNo - segno;
+                walSegment = strLstGet(segmentList, i);
+            }
+        }
+        else
+        {
+            if (segno - fileSegNo < segnoDiff && fileSegNo < segno)
+            {
+                segnoDiff = segno - fileSegNo;
+                walSegment = strLstGet(segmentList, i);
+            }
+        }
+    }
+
+    // current file is oldest/newest
+    if (segnoDiff == UINT64_MAX)
+    {
+        return NULL;
+    }
+
+    const bool compressible =
+        this->archiveInfo->cipherType == cipherTypeNone && compressTypeFromName(this->archiveInfo->file) == compressTypeNone;
+
+    const StorageRead *const storageRead = storageNewReadP(
+        storageRepoIdx(0),
+        strNewFmt(STORAGE_REPO_ARCHIVE "/%s/%s", strZ(this->archiveInfo->archiveId), strZ(walSegment)),
+        .compressible = compressible);
+
+    buildArchiveGetPipeLine(ioReadFilterGroup(storageReadIo(storageRead)), this->archiveInfo);
+    return storageRead;
+}
+
+static bool
+readBeginOfRecord(WalFilterState *const this)
+{
+    bool result = false;
+    MEM_CONTEXT_TEMP_BEGIN();
+
+    const StorageRead *const storageRead = getNearWal(this, false);
+
+    if (storageRead == NULL)
+    {
+        goto end;
+    }
+
+    ioReadOpen(storageReadIo(storageRead));
+    this->isBegin = true;
+    this->inputOffset = 0;
+    this->pageOffset = 0;
+
+    Buffer *const buffer = bufNew(this->walPageSize);
+    size_t size = ioRead(storageReadIo(storageRead), buffer);
+    bufUsedSet(buffer, size);
+
+    while (true)
+    {
+        if (readRecord(this, buffer) == ReadRecordNeedBuffer)
+        {
```

Comment:
Can you rewrite this the same way as done in the `getEndOfRecord` function?

Reply:
No. Since in the case of `getEndOfRecord`, we are reading one record. And in this function we read all the records.

Reply:
Exiting the loop doesn't seem very reliable, maybe loop until the end of the file is reached?

Reply:
done